skglm/utils.py

```diff
@@ -254,3 +254,55 @@ def check_group_compatible(obj):
                 f"'{obj_name}' is not block-separable. "
                 f"Missing '{attr}' attribute."
             )
+
+
+class AndersonAcceleration:
+    """Abstraction of Anderson Acceleration.
+
+    Extrapolate the asymptotic VAR ``w`` and ``Xw``
+    based on ``K`` previous iterations.
+
+    Parameters
+    ----------
+    K : int
+        Number of previous iterates to consider for extrapolation.
+
+    n_features : int
+        Number of elements in ``w``.
+
+    n_samples: int
+        Number of elements in ``Xw``.
+    """
+
+    def __init__(self, K, n_samples, n_features):
+        self.K = K
+        self.n_samples, self.n_features = n_samples, n_features
+
+        self.current_iter = 0
+        self.arr_w = np.zeros(shape=(n_features, K+1))
+        self.arr_Xw = np.zeros(shape=(n_samples, K+1))
+
+    def extrapolate(self, w, Xw):
+        """Inplace update of ``w`` and ``Xw``."""
+        if self.current_iter <= self.K:
+            self.arr_w[:, self.current_iter] = w.copy()
+            self.arr_Xw[:, self.current_iter] = Xw.copy()
+            self.current_iter += 1
+            return
+
+        # compute residuals
+        U = self.arr_w[:, 1:] - self.arr_w[:, :-1]
+
+        # compute extrapolation coefs
+        ones_K = np.ones(self.K)
+        try:
+            inv_UTU_ones = np.linalg.solve(U.T @ U, ones_K)
+        except np.linalg.LinAlgError:
+            return
+        finally:
+            self.current_iter = 0
+
+        # extrapolate
+        C = inv_UTU_ones / (ones_K @ inv_UTU_ones)
+        w[:] = self.arr_w[:, 1:] @ C
+        Xw[:] = self.arr_Xw[:, 1:] @ C
```

Comment:
add a comment saying that due to floating point errors, this may cause w and Xw to disagree